packages/vitest/src/integrations/snapshot/port/utils.ts

```diff
@@ -157,3 +157,55 @@ export async function saveSnapshotFile(
     'utf-8',
   )
 }
+
+function isObject(item: unknown): boolean {
+  return item != null && typeof item === 'object' && !Array.isArray(item)
+}
+
+function deepMergeArray(target: Array<any>, source: Array<any>) {
+  const mergedOutput = Array.from(target)
+
+  source.forEach((sourceElement, index) => {
+    const targetElement = mergedOutput[index]
+
+    if (Array.isArray(target[index])) {
+      mergedOutput[index] = deepMergeArray(target[index], sourceElement)
+    }
+    else if (isObject(targetElement)) {
+      mergedOutput[index] = deepMerge(target[index], sourceElement)
+    }
+    else {
+      // Source does not exist in target or target is primitive and cannot be deep merged
+      mergedOutput[index] = sourceElement
+    }
+  })
+
+  return mergedOutput
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function deepMerge(target: any, source: any): any {
```

Comment:
we already have `deepMerge` but it behaves very wierdly - id doesn't support arrays and checks for the type of object with `toString.call(val)`